### PR TITLE
Ordinary calling default with regions

### PR DESCRIPTION
**Pull request by 7heo**
Makes boto work over HTTPS with buckets containing dots in their name (see #2836)

### CODE REVIEW

**Comment by vlcinsky** [comment]
@7heo Unfortunatelly, it does not work for me.

Installed virtualbox with debian-jessie having Python 2.7.9

```
vagrant@jessie-amd64:~$ lsb_release -a
No LSB modules are available.
Distributor ID: Debian
Description:    Debian GNU/Linux 8.0 (jessie)
Release:        8.0
Codename:       jessie
```
Having python 2.7.9:
```
vagrant@jessie-amd64:~$ python
Python 2.7.9 (default, Mar  1 2015, 12:57:24) 
[GCC 4.9.2] on linux2
Type "help", "copyright", "credits" or "license" for more information.
>>> exit()
```

Running in fresh virtualenv, installing boto from your latest commit:
```
(botodoto)vagrant@jessie-amd64:~/sandbox/botodoto/boto$ git log -1
commit 0b1eb863bc0e8bccefd9a7475b01b502bb1014a6
Author: 7heo <7heo@mail.com>
Date:   Fri May 8 01:19:43 2015 +0200

    Attempt to fix the broken tests
(botodoto)vagrant@jessie-amd64:~/sandbox/botodoto/boto$ pip freeze
-e git+https://github.com/7heo/boto.git@0b1eb863bc0e8bccefd9a7475b01b502bb1014a6#egg=boto-ordinary-calling-default-with-regions
ipython==3.1.0
```
Using `.boto` configuration file in my home folder
```
vagrant@jessie-amd64:~$ cat ~/.boto
[Credentials]
aws_access_key_id = <some misty weather here>
aws_secret_access_key = <here is heavy raining>
# region = eu-west-1
```

Without having region specified in my `.boto`, it fails with Moved permanently:
```
In [1]: import boto

In [2]: con = boto.connect_s3()

In [3]: bucket_name = "cz-base.dp.ce-traffic.com"

In [4]: bucket = con.get_bucket(bucket_name)
---------------------------------------------------------------------------
S3ResponseError                           Traceback (most recent call last)
<ipython-input-4-6300deb73365> in <module>()
----> 1 bucket = con.get_bucket(bucket_name)

/home/vagrant/sandbox/botodoto/boto/boto/s3/connection.pyc in get_bucket(self, bucket_name, validate, headers, location)
    547         self.current_bucket_location = (location if location in Location else None)
    548         if validate:
--> 549             return self.head_bucket(bucket_name, headers=headers)
    550         else:
    551             return self.bucket_class(self, bucket_name)

/home/vagrant/sandbox/botodoto/boto/boto/s3/connection.pyc in head_bucket(self, bucket_name, headers)
    594         else:
    595             raise self.provider.storage_response_error(
--> 596                 response.status, response.reason, body)
    597 
    598     def lookup(self, bucket_name, validate=True, headers=None):

S3ResponseError: S3ResponseError: 301 Moved Permanently
```

When I enable region in my `.boto` to `eu-west-1` (where are my buckets really present), it ends up about the same:
```
In [1]: import boto

In [2]: con = boto.connect_s3()

In [3]: bucket_name = "cz-base.dp.ce-traffic.com"

In [4]: bucket = con.get_bucket(bucket_name)
---------------------------------------------------------------------------
S3ResponseError                           Traceback (most recent call last)
<ipython-input-4-6300deb73365> in <module>()
----> 1 bucket = con.get_bucket(bucket_name)

/home/vagrant/sandbox/botodoto/boto/boto/s3/connection.pyc in get_bucket(self, bucket_name, validate, headers, location)
    547         self.current_bucket_location = (location if location in Location else None)
    548         if validate:
--> 549             return self.head_bucket(bucket_name, headers=headers)
    550         else:
    551             return self.bucket_class(self, bucket_name)

/home/vagrant/sandbox/botodoto/boto/boto/s3/connection.pyc in head_bucket(self, bucket_name, headers)
    594         else:
    595             raise self.provider.storage_response_error(
--> 596                 response.status, response.reason, body)
    597 
    598     def lookup(self, bucket_name, validate=True, headers=None):

S3ResponseError: S3ResponseError: 301 Moved Permanently
```

When I create connection for specific region, it then stays stale for a while a step later at `bucket.get_key()`:
```
In [1]: import boto.s3

In [2]: con = boto.s3.connect_to_region("eu-west-1")

In [3]: bucket_name = "cz-base.dp.ce-traffic.com"

In [4]: bucket = con.get_bucket(bucket_name)

In [5]: bucket
Out[5]: <Bucket: cz-base.dp.ce-traffic.com>

In [6]: key_name = "region/cz/EventService.xml"

In [7]: key = bucket.get_key(key_name)
---------------------------------------------------------------------------
gaierror                                  Traceback (most recent call last)
<ipython-input-7-5bd0ce514b3e> in <module>()
----> 1 key = bucket.get_key(key_name)

/home/vagrant/sandbox/botodoto/boto/boto/s3/bucket.pyc in get_key(self, key_name, headers, version_id, response_headers, validate)
    190                 query_args_l.append('%s=%s' % (rk, urllib.parse.quote(rv)))
    191 
--> 192         key, resp = self._get_key_internal(key_name, headers, query_args_l)
    193         return key
    194 

/home/vagrant/sandbox/botodoto/boto/boto/s3/bucket.pyc in _get_key_internal(self, key_name, headers, query_args_l)
    197         response = self.connection.make_request('HEAD', self.name, key_name,
    198                                                 headers=headers,
--> 199                                                 query_args=query_args)
    200         response.read()
    201         # Allow any success status (2xx) - for example this lets us

/home/vagrant/sandbox/botodoto/boto/boto/s3/connection.pyc in make_request(self, method, bucket, key, headers, data, query_args, sender, override_num_retries, retry_handler, needs_location)
    716             data, host, auth_path, sender,
    717             override_num_retries=override_num_retries,
--> 718             retry_handler=retry_handler
    719         )

/home/vagrant/sandbox/botodoto/boto/boto/connection.pyc in make_request(self, method, path, headers, data, host, auth_path, sender, override_num_retries, params, retry_handler)
   1069                                                     params, headers, data, host)
   1070         return self._mexe(http_request, sender, override_num_retries,
-> 1071                           retry_handler=retry_handler)
   1072 
   1073     def close(self):

/home/vagrant/sandbox/botodoto/boto/boto/connection.pyc in _mexe(self, request, sender, override_num_retries, retry_handler)
   1028             raise BotoServerError(response.status, response.reason, body)
   1029         elif ex:
-> 1030             raise ex
   1031         else:
   1032             msg = 'Please report this exception as a Boto Issue!'

gaierror: [Errno -2] Name or service not known
```

Checking region of the bucket, I see, it is really in eu-west-1:
```
In [8]: bucket.get_location()
Out[8]: u'eu-west-1'
```

**Comment by 7heo** [comment]
Yeah head buckets cause problems with my patch. I noticed the failures before going to bed, but I was too tired so I didn't fix them yet. It's on today's agenda.

Long story short, I changed too much of the default behavior of make_request (which is doing something different by *default* instead on *on request*). I think the errors you see as well as the CI failure come from that. I'll change that and try again. More news as I try things ;)

PS: I don't have much to test with locally, so the only way I have to test is the CI as of now... Which means slower, more painful testing (no access to the remote for interactive tests). Hopefully that is going to suffice.

**Comment by 7heo** [comment]
Alright, I've identified the reason for the breakage: In my eagerness to PR, I have left my modifications without proper checks, since it was working in my local tests. Fixing that now.

**Comment by 7heo** [comment]
Alright that PR is too big, I'm going to test things individually.

**Comment by 7heo** [comment]
PR replaced by https://github.com/boto/boto/pull/3163

File: boto/s3/connection.py [hunk 1]
```diff
@@ -135,17 +139,36 @@ def build_url_base(self, connection, protocol, server, bucket, key=''):
...
+    # Allows to access unexisting members of Location without having an
+    # instance of Location.
+    def __init__(self, name, bases, attrs):
+        super(MetaLocation, self).__init__(name, bases, attrs)
+        self.__regions = {
+            # For up to date information, see:
+            # http://docs.aws.amazon.com/general/latest/gr/rande.html#s3_region
+            'DEFAULT':      '',  # US Classic Region
+            'EU':           'eu-west-1',
+            'EUWest':       'eu-west-1',
```
Comment (malexejev) [inline boto/s3/connection.py]: Seems like a bug - two "eu-west-1" on after another

Comment (7heo) [inline boto/s3/connection.py]: It's using different keys, with a similar value. That's nothing like a bug, it's the desired behavior.

**Pull request closed without merging**